docs/architecture/adr-014-versioned-namespaces.md

```diff
@@ -0,0 +1,192 @@
+
+- 2023/2/14: initial draft
+
+## Context
+
+The specs-staging branch contains the current schema for a [sparse share](https://github.com/celestiaorg/celestia-app/blob/0dc7d17636efa535efca42af58229ee0c3c21261/specs/src/specs/data_structures.md#sparse-share). A brief reminder of the schema for the first blob share in a sequence:
+
+![first-sparse-share](./assets/first-sparse-share.svg)
+
+| Field | Number of bytes | Description |
+| --- | --- | --- |
+| namespace id | 8 | namespace ID of the share |
+| info byte | 1 | the first 7 bits determine the share version. The last 1 bit determines if this is the first share in a sequence |
+| sequence length | 4 | the number of bytes in the sequence encoded as a big-endian uint32 |
+| msg1 | 499 | blob1's raw data |
+
+```go
+currentBlobShare := []byte{
+  1, 2, 3, 4, 5, 6, 7, 8, // namespace ID
+  1,          // info byte (sequence start indicator = true)
+  0, 0, 0, 3, // sequence length
+  1, 2, 3, // blob data
+  // 0 padding until share is full
+ }
+```
+
+The current schema poses a challenge for the following scenarios:
+
+1. Consider increasing the namespace id size [celestia-app#1308](https://github.com/celestiaorg/celestia-app/issues/1308)
+    - Since there is no prefix to the namespace ID, it isn't possible to distinguish between a share with an 8 byte namespace ID and a larger namespace ID (say 16 bytes).
+2. Deterministic namespace ID based on blob content [celestia-app#1377](https://github.com/celestiaorg/celestia-app/issues/1377)
+    - Assuming we use SHA-256 and the deterministic namespace ID has a length of 32 bytes, it isn't possible to distinguish between an 8 byte namespace ID and a 32 byte namespace ID.
+3. Changes to the non-interactive default rules that don't break backwards compatibility with existing namespaces [celestia-app#1282](https://github.com/celestiaorg/celestia-app/issues/1282), [celestia-app#1161](https://github.com/celestiaorg/celestia-app/pull/1161)
+    - After mainnet launch, if we want to change the non-interactive default rules but retain the previous non-interactive default rules for backwards compatibility, it isn't possible to differentiate the namespaces that want to use the old rules vs the new rules.
+
+## Proposal
+
+An approach that addresses these issues is to prefix the namespace ID with version metadata.
+
+### Option A: `Namespace Version | Namespace ID`
+
+| Field | Number of bytes | Description |
+| --- | --- | --- |
+| Namespace Version | 1 | the version of the namespace ID |
+| Namespace ID | 8 if Namespace Version=0, 32 if Namespace Version=1 | namespace ID of the share |
+
+For example, consider the scenario where at mainnet launch namespace IDs have a length of 8 bytes. In this scenario, the only supported namespace ID is `0`. At some point in the future,  we want to increase the length of namespace IDs from 8 bytes to 32 bytes. We can expand the range of available namespaces to include namespaces that start with a leading `0` or `1` byte.
```

Comment:
Q: I might have misunderstood that on the call but didn't we say that despite the namespace version increasing the namespace size would be a breaking/non-backwards compatible change where the version alone will not help (because it would also impact the roots etc). If that is the case (and I understood it correctly) I'm in favour with using a proper use-case here instead.

Reply:
You understood correctly. I'll update this section with a different use case.

Reply:
Resolved by https://github.com/celestiaorg/celestia-app/pull/1382/commits/e579db3ac970fe7a7097d48be51c43f6e1e3d2bd